Consolidate build instructions into BUILD.md

mfdorst opened the pull request:
Remove "quick start guide" from README - replace it with link to
BUILD.md

Add instructions for applying print-spilling-info patch, among other things

I chose `BUILD.md` because I feel like "install" is the wrong word, but I'm not attached to it.

Closes #7

mfdorst commented:
I'd also like to make a few changes regarding ninja, but wanted to ask about them first.

1. Ninja can be installed [using apt]. Can we switch from the `wget` installation to using `apt`?
This would remove a whole section from the BUILD doc, making it easier to follow.
Instead we can add it to the earlier instruction: `apt install git cmake ninja`

2. According to [Ninja's documentation]:
> Ninja defaults to running commands in parallel anyway, so typically you don’t need to pass -j
So I propose changing the command from `ninja -j32` to `ninja`.

[using apt]: https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages
[ninja's documentation]: https://ninja-build.org/manual.html#_running_ninja

kerbowa commented:
> 1. Ninja can be installed [using apt](https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages). Can we switch from the `wget` installation to using `apt`?
>    This would remove a whole section from the BUILD doc, making it easier to follow.
>    Instead we can add it to the earlier instruction: `apt install git cmake ninja`

I thought that in Ubuntu 16.04 the version of ninja in the official package was too outdated to work, but for 18.04 it's fine. I kind of wanted to have some set of instructions that would work well with copy pasting the commands, but maybe it is over complicating things. Maybe instead of using Ninja we should just suggest to use an alternative linker like gold or lld.

> 2. According to [Ninja's documentation](https://ninja-build.org/manual.html#_running_ninja):
> 
> > Ninja defaults to running commands in parallel anyway, so typically you don’t need to pass -j
> > So I propose changing the command from `ninja -j32` to `ninja`.

Okay.

mfdorst commented on BUILD.md:
````diff
@@ -183,7 +183,7 @@ cmake -DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Debug '-DLLVM_TARGETS_TO
 ```
 
 ```
-make -j 32
````
The original instructions just have `make`. I have been doing `make -j<number>` because I thought that made it multithreaded, but upon checking the documentation it actually just _limits_ the number of simultaneous jobs to that number. Make will automatically figure out the right number of jobs for you if you leave it off.

mfdorst replied:
It's for the same reason that I'm proposing to remove `-32` from the suggested `ninja` command. You're not telling it to use 32 threads, you're telling it _not_ to use _more_ than 32 threads, which seems pointless to me.

kerbowa replied:
Well that's not true for make, although I'm not sure of the details with ninja.

The point of suggesting to use ninja in the first place is that you can limit the number of link jobs. With 'make -j32' you can run out of memory doing a debug build of LLVM. This shouldn't happen with just 'make'.

mfdorst replied:
Are we on the same page here? Currently the directions I wrote are to just use `make` and `ninja` with no `-j` flag. I'm not sure if you're wanting to keep one or both of them as `-j32`?

mfdorst commented:
Oh I didn't realize that. I think you're right then, it's better to just have one command that works on all versions. I think I'll change the `MacOS` directions to use `brew install ninja` though.

mfdorst commented:
I changed my mind. I've been erring on the side of completeness in this document, so I added both options. I'm not too worried about the document getting too long, because the table of contents should make it easy to find where you left off.

kerbowa commented:
LGTM
